massa-execution/src/worker.rs

```diff
@@ -240,121 +239,194 @@ impl ExecutionWorker {
+        let new_css_final_blocks = finalized_blocks.into_iter().filter_map(|(b_id, b)| {
+            if b.header.content.slot <= self.last_active_slot {
+                // eliminate blocks that are not from a stricly later slot than the current latest SCE-final one
+                // this is an optimization
+                return None;
             }
-            loop {
-                let next_final_slot = self.last_final_slot.get_next_slot(self.cfg.thread_count)?;
-                if next_final_slot == block_slot {
-                    self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
-                        slot: next_final_slot,
-                        block: Some((b_id, block.clone())),
-                    }));
-                    self.last_final_slot = next_final_slot;
-                    break;
-                } else if next_final_slot < max_thread_slot[next_final_slot.thread as usize] {
-                    self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
-                        slot: next_final_slot,
-                        block: Some((b_id, block.clone())),
-                    }));
-                    self.last_final_slot = next_final_slot;
-                } else {
-                    self.ordered_pending_css_final_blocks.push((b_id, block));
-                    break;
+            Some((b.header.content.slot, (b_id, b)))
+        });
+        self.pending_css_final_blocks.extend(new_css_final_blocks);
+
+        if let Some(max_css_final_slot) = self
+            .pending_css_final_blocks
+            .last_key_value()
+            .map(|(s, _v)| *s)
+        {
+            // iterate over every slot S starting from `last_final_slot.get_next_slot()` up to the latest slot in `pending_css_final_blocks` (included)
+            let mut s = self.last_final_slot.get_next_slot(self.cfg.thread_count)?;
+            while s <= max_css_final_slot {
+                match self
+                    .pending_css_final_blocks
+                    .first_key_value()
+                    .map(|(s, _v)| *s)
+                {
+                    // there is a block B at slot S in `pending_css_final_blocks`:
+                    Some(b_slot) if b_slot == s => {
+                        // remove B from `pending_css_final_blocks`
+                        // cannot panic, checked above
+                        let (_s, (b_id, b)) = self
+                            .pending_css_final_blocks
+                            .pop_first()
+                            .expect("pending_css_final_blocks wa unexpectedly empty");
+                        // call the VM to execute the SCE-final block B at slot S
+                        self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
+                            slot: s,
+                            block: Some((b_id, b)),
+                        }));
+                        // set `last_active_slot = last_final_slot = S`
```

Comment:
why that comment ?

Reply:
see Greg's comments on https://github.com/massalabs/massa/pull/2019

=> match comments 1-to-1 with the spec for linear reading

Reply:
not sure it helps with linear reading in this case

Reply:
removed